pkg/solana/chainreader/chain_reader.go

```diff
@@ -91,8 +93,35 @@ func (s *SolanaChainReaderService) GetLatestValue(ctx context.Context, contractN
 		return err
 	}
 
+	if len(bindings) > 1 {
+		// might go for some guardrails when dealing with multiple bindings
+		// the returnVal should be compatible with multiple passes by the codec decoder
+		// this should only apply to types struct{} and map[any]any
+		tReturnVal := reflect.TypeOf(returnVal)
+		if tReturnVal.Kind() == reflect.Pointer {
+			tReturnVal = reflect.Indirect(reflect.ValueOf(returnVal)).Type()
+		}
+
+		switch tReturnVal.Kind() {
+		case reflect.Struct, reflect.Map:
+		default:
+			return fmt.Errorf("%w: multiple bindings is only supported for struct and map", types.ErrInvalidType)
+		}
+
+		// for multiple bindings, preload the remote data in parallel
+		for _, binding := range bindings {
+			go func(ctx context.Context, rb readBinding) {
```

Comment:
It looks like the lifecyles of these goroutines needs to be managed carefully, since we can short circuit down below and return error before waiting for all the bindings. The simplest thing to do would be to cancel a context and block on a local wait group before returning. If that introduces weird latency, then they could still be "abandoned" from this method as long as `Close()`ing this `SolanaChainReaderService` waits for them. That would be more complicated to wire up though.

Reply:
I decided to wrap the incoming context with a cancel context and add a wait group. Then I cancel the context on exit and wait for the wait group to complete. The only caveat is that if the remote call doesn't respect the context and exit, the call to `GetLatestValue` on the service could hang.

Also I wired up another wait group on the service `Close` function. The only issue there is that it doesn't call cancel to any existing processes.